File: backend/scala-seed-project/src/services/RoleService.scala
// src/services/RoleService.scala

import cats.effect.IO

object RoleService {

  def createRole(name: String, permissions: List[String]): IO[Role] = {
    for {
      _ <- IO.raiseError(new Exception("Invalid input: name and permissions must be provided"))
      existingRoleOpt <- RoleRepository.findByName(name)
      _ <- IO.fromOption(existingRoleOpt)(new Exception(s"Role with name $name already exists"))
      newRole <- RoleRepository.create(Role(None, name, permissions))
    } yield newRole
  }

  def deleteRoles(roleIds: List[Int]): IO[Int] = {
    RoleRepository.delete(roleIds)
  }

  def changeRoleName(roleId: Int, newName: String): IO[Option[Role]] = {
    RoleRepository.changeName(roleId, newName)
  }

  def changeRolePermissions(roleId: Int, newPermissions: List[String]): IO[Option[Role]] = {
    RoleRepository.changePermissions(roleId, newPermissions)
  }

  def getAllRoles(): IO[List[Role]] = {
    RoleRepository.findAll()
  }
}

File: backend/node_version/src/services/roleService.ts
import Role from '../models/roleModel'; 

export const createRole = async (name: string, permissions: string[]) => {
  if (!name || !Array.isArray(permissions)) {
    throw new Error('Invalid input: name and permissions must be provided');
  }

  const existingRole = await Role.findOne({ where: { name } });
  if (existingRole) {
    throw new Error(`Role with name ${name} already exists`);
  }

  return Role.create({ name, permissions });
};

export const deleteRoles = async (roleIds: string[]) => {
  const deletedRoles = await Role.destroy({
    where: {
      id: roleIds, // Use the IDs to find roles
    },
  });

  if (!deletedRoles) {
    throw new Error('No roles found with the provided IDs.');
  }

  return deletedRoles;
};


export const changeRoleName = async (roleId: number, newName: string) => {
  const role = await Role.findByPk(roleId);
  if (!role) throw new Error('Role not found');
  role.name = newName;
  await role.save();
  return role;
};

export const changeRolePermissions = async (roleId: number, newPermissions: string[]) => {
  const role = await Role.findByPk(roleId);
  if (!role) throw new Error('Role not found');
  role.permissions = newPermissions;
  await role.save();
  return role;
};

export const getAllRoles = async () => {
  return Role.findAll();
};
